tests/dup/programs/dup/src/lib.rs

```diff
@@ -0,0 +1,114 @@
+        Ok(())
+    }
+
+    pub fn without_dup_constraint_composite(_ctx: Context<WithoutDupConstraintComposite>) -> ProgramResult {
+        Ok(())
+    }
+
+    pub fn with_missing_dup_constraints_three_accounts(_ctx: Context<WithMissingDupConstraintThreeAccounts>) -> ProgramResult {
+        Ok(())
+    }
+
+    pub fn with_dup_constraints_three_accounts(_ctx: Context<WithDupConstraintsThreeAccounts>) -> ProgramResult {
+        Ok(())
+    }
+
+    pub fn with_missing_dup_constraint_double_three_accounts(_ctx: Context<WithMissingDupConstraintDoubleThreeAccounts>) -> ProgramResult {
+        Ok(())
+    }
+
+    pub fn without_dup_constraint_double_three_accounts_all_immutable(_ctx: Context<WithoutDupConstraintDoubleThreeAccountsAllImmutable>) -> ProgramResult {
+        Ok(())
+    }
+}
+
+#[derive(Accounts)]
+pub struct WithDupConstraint<'info> {
+    pub account1: Signer<'info>,
+    #[account(dup = account1)]
+    pub account2: SystemAccount<'info>,
+}
+
+#[derive(Accounts)]
+pub struct WithDupConstraintComposite<'info> {
+    #[account(dup = child.child_account)]
+    pub account1: SystemAccount<'info>,
+    pub child: Child<'info>
+}
+
+#[derive(Accounts)]
+pub struct Child<'info> {
+    pub child_account: SystemAccount<'info>
+}
+
+#[derive(Accounts)]
+pub struct WithoutDupConstraint<'info> {
+    pub account1: SystemAccount<'info>,
+    #[account(mut)]
+    pub account2: SystemAccount<'info>,
+}
+
+#[derive(Accounts)]
+pub struct WithoutDupConstraintComposite<'info> {
+    pub child: Child<'info>,
+    #[account(mut)]
+    pub account1: SystemAccount<'info>,
+}
+
+#[derive(Accounts)]
+pub struct WithMissingDupConstraintThreeAccounts<'info> {
+    pub account1: SystemAccount<'info>,
+    #[account(dup = account1, mut)]
+    pub account2: SystemAccount<'info>,
+    pub account3: SystemAccount<'info>,
+}
+
+#[derive(Accounts)]
+pub struct WithDupConstraintsThreeAccounts<'info> {
+    pub account1: SystemAccount<'info>,
+    #[account(dup = account1)]
+    pub account2: SystemAccount<'info>,
+    #[account(mut, dup = account1)]
+    pub account3: SystemAccount<'info>,
+}
+
+#[derive(Accounts)]
+pub struct WithMissingDupConstraintDoubleThreeAccounts<'info> {
+    pub account1: SystemAccount<'info>,
+    #[account(dup = account1)]
+    pub account2: SystemAccount<'info>,
+    #[account(dup = account1)]
+    pub account3: SystemAccount<'info>,
+    pub account4: SystemAccount<'info>,
+    #[account(dup = account4)]
+    pub account5: SystemAccount<'info>,
+    #[account(mut)]
+    pub account6: SystemAccount<'info>,
+}
+
+#[derive(Accounts)]
+pub struct WithoutDupConstraintDoubleThreeAccountsAllImmutable<'info> {
+    pub account1: SystemAccount<'info>,
+    pub account2: SystemAccount<'info>,
+    pub account3: SystemAccount<'info>,
+    pub account4: SystemAccount<'info>,
+    pub account5: SystemAccount<'info>,
+    pub account6: SystemAccount<'info>,
+}
```

Comment:
What happens if I have a struct like this

```rust
#[derive(Accounts)]
pub struct MyAccounts {
  pub composite_a: CompositeA,
  pub composite_b: CompositeB,
}
```

Where there are duplicates within `composite_a` and `composite_b`?

Reply:
the macro on MyAccount cannot look inside CompositeA or CompositeB so it cannot enforce any relationships between accounts in composite_a and composite_b.

the two composite structs will each check whether the accounts inside them are duplicates

Reply:
if that is how these macros work, then it could still be possible to compare composite_a and composite_b. have the macro implement a function (e.g. `fields`) for each Accounts struct that just returns its accountinfos. then the macro can - when it sees a composite field - call `fields` on that struct and expand to a loop that checks the current value to each field in that struct. This would require hardcoding which fields on a struct are `dup` in the `fields` function so it returns Iterable<(is_dup, accountinfo)>

Reply:
The `fields` function sounds like nice way to cover this case.

Reply:
there's one more thing to think about then

assuming that we dont want Accounts struct to have constraints upwards (only into their own fields) to keep composability, there will be no way of using the `maybe_dup(x)` constraint from a composite field upwards to its parent.

this is also the case with all other constraints tho so dont think it is a problem?

Reply:
The constraint shouldn't necessarily apply upward, but if there's a dup in any sub (or sibling) accounts struct, it should be caught.